modules/howtos/examples/Import.java

```diff
@@ -0,0 +1,411 @@
+    importer.importJSONL();
+    importer.importCSV_batch();
+    importer.importTSV_batch();
+    importer.importJSON_batch();
+    importer.importJSONL_batch();
+  }
+  
+  // NOTE: non-generic Map, because the readers produce slightly different outputs:
+  //   CSV: Map<String,String>
+  //   JSON: Map<String,Object>
+  // and as generics annotations are lost at runtime we can't have
+  // a single overloaded method to deal with those.
+  //
+  // In real code, you might choose to cast from Object to Map<String,Object>
+  // or define multiple methods, but those confuse the example a little.
+  
+  // tag::preprocess[]
+  public JsonDocument preprocess(Map row) {
+    Map value = new HashMap(row);
+
+    // define the KEY
+    String key = value.get("type") + "_" + value.get("id");
+
+    // do any additional processing
+    value.put("importer", "Java SDK");
+    
+    return new JsonDocument(key, JsonObject.from(value));
+  }
+  // end::preprocess[]
+  
+  // tag::upsertDocument[]
+  public void upsertRow(Map row) {
+    
+    JsonDocument doc = preprocess(row);
+    
+    String key = doc.getId();
+    Object value = doc.getContent();
+    
+    // upsert the document
+    collection.upsert(key, value);
+    
+    // any required logging
+    System.out.println(key);
+    System.out.println(value);
+  }
+  // end::upsertDocument[]
+  
+  
+  // tag::importCSV[]
+  public void importCSV() {
+    try {
+      CSVReaderHeaderAware csv = new CSVReaderHeaderAware(
+        new FileReader("howtos/examples/import.csv"));
+      
+      Map<String, String> row;
+      while ((row = csv.readMap()) != null) {
+        upsertRow(row);
+      }
+    }
+    catch (java.io.FileNotFoundException e) {
+      System.out.println("handle FileNotFoundException...");
+    }
+    catch (java.io.IOException e) {
+      System.out.println("handle IOException...");
+    }
+    catch (com.opencsv.exceptions.CsvValidationException e) {
+      System.out.println("handle CsvValidationException...");
+    }
+  }
+  // end::importCSV[]
+  
+  
+  // tag::importCSV_batch[]
+  public void importCSV_batch() {
+    // tag::omit[]
+    System.out.println("importCSV_batch");
+    // end::omit[]
+    
+    Flux<Map<String,String>> rows = Flux.generate(
+      
+      () -> new CSVReaderHeaderAware(
+        new FileReader("howtos/examples/import.csv")),
+      
+      (state, sink) -> {
+        try {
+          Map<String,String> row = state.readMap();
+          if (row == null) { sink.complete(); }
+          else { sink.next(row); }
+          return state;
+        }
+        catch (IOException e) { throw new Error("IOException"); }
+        catch (com.opencsv.exceptions.CsvValidationException e) { throw new Error("IOException"); }
+      });
+
+    Flux<MutationResult> results = 
+      rows
+      .map(row -> preprocess(row))
+      .flatMap(doc -> reactiveCollection.upsert(doc.getId(), doc.getContent()));
+      
+    results.subscribe(System.out::println);
```

Comment:
This code is upserting the documents twice, since `subscribe` and `blockLast` both subscribe to the flux (triggering its execution).

Recommendation is to remove the `subscribe`. If you want to print each mutation result, one way would be to chain `.doOnNext(System.out::println)` after `flatMap`.

Reply:
ah, I thought as it's push based, the code would be run once and the *result* sent to all subscribers?
I'll have to review my mental model... thanks, will look at .doOnNext!